src/AutoRest.CSharp/Common/Output/PostProcessing/PostProcessor.cs

```diff
@@ -0,0 +1,291 @@
+using AutoRest.CSharp.Output.Models.Types;
+using AutoRest.CSharp.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace AutoRest.CSharp.Common.Output.PostProcessing;
+
+internal abstract class PostProcessor
+{
+    public PostProcessor()
+    {
+    }
+
+    protected record TypeSymbols(HashSet<INamedTypeSymbol> DeclaredSymbols, Dictionary<INamedTypeSymbol, HashSet<BaseTypeDeclarationSyntax>> DeclaredNodesCache, Dictionary<Document, HashSet<INamedTypeSymbol>> DocumentsCache);
+
+    /// <summary>
+    /// This method reads the project, returns the types defined in it and build symbol caches to acceralate the calculation
+    /// By default, the types defined in shared documents are not included. Please override <see cref="ShouldIncludeDocument(Document)"/> to tweak this behaviour
+    /// </summary>
+    /// <param name="compilation">The <see cref="Compilation"/> of the <paramref name="project"/> </param>
+    /// <param name="project">The project to extract type symbols from</param>
+    /// <param name="publicOnly">If <paramref name="publicOnly"/> is true, only public types will be included. If <paramref name="project"/> is false, all types will be included </param>
+    /// <returns>A instance of <see cref="TypeSymbols"/> which includes the information of the declared symbols of the given accessibility, along with some useful cache that is useful in this class. </returns>
+    protected async Task<TypeSymbols> GetTypeSymbolsAsync(Compilation compilation, Project project, bool publicOnly = true)
+    {
+        var result = new HashSet<INamedTypeSymbol>(SymbolEqualityComparer.Default);
+        var declarationCache = new Dictionary<INamedTypeSymbol, HashSet<BaseTypeDeclarationSyntax>>(SymbolEqualityComparer.Default);
+        var documentCache = new Dictionary<Document, HashSet<INamedTypeSymbol>>();
+        foreach (var document in project.Documents)
+        {
+            if (ShouldIncludeDocument(document))
+            {
+                var root = await document.GetSyntaxRootAsync();
+                if (root == null)
+                    continue;
+
+                var semanticModel = compilation.GetSemanticModel(root.SyntaxTree);
+
+                foreach (var typeDeclaration in root.DescendantNodes().OfType<BaseTypeDeclarationSyntax>())
+                {
+                    var symbol = semanticModel.GetDeclaredSymbol(typeDeclaration);
+                    if (symbol == null)
+                        continue;
+                    if (publicOnly && symbol.DeclaredAccessibility != Accessibility.Public)
+                        continue;
+                    result.Add(symbol);
+                    declarationCache.AddInList(symbol, typeDeclaration);
+                    documentCache.AddInList(document, symbol, () => new HashSet<INamedTypeSymbol>(SymbolEqualityComparer.Default));
```

Comment:
Correct me if I'm wrong:
- `result` contains exactly the same symbols as `declarationCache.Keys`
- `documentCache` may contain only one type symbol per document, otherwise it is ignored in `ProcessSymbolAsync`
- It is assumed that we can't have more than one top level type in the file

Reply:
1. Correct
2. No. In `ProcessSymbolAsync` when we are finding references of one symbol A, we can only get the text span and document of where it is used. But in the map we need the type symbol that is referencing symbol A. I did not find a handy API in roslyn to do that. Therefore considering the reference information contains the "document", I keep a cache of "what symbols a document could hold". When we want to get a symbol from the document, and the document only contains one, we just use that symbol without any calculation. If the document contains multiple, we still have to determine which symbol in the document is referencing symbol A (this case will be very rare in our generated code)
3. No, we did not make that assumption and it is also the reason why I iterate over all the type declaration nodes under one root